### PR TITLE
Test branch for paper

### LINKED ISSUE #217: Fix Hierarchical Clustering
## Issue description
Hierarchical Clustering has some issues. Will update this Issue's TODOs as I identify and fix them. Closely related to #201 and I'm working directly on `test-branch-for-paper` for this

## The Problems
- [ ] Saving and reloading a run with Hierarchical Agglomerative Clustering (HAC) doenst work as the `model` data is saved as `!!python/object:sklearn.cluster._agglomerative.AgglomerativeClustering` and there is no constructor to reconstruct the data from that
- [ ] HAC does not produce any plots. Getting the HAC step output from sklearn converted to something that the existing Clustergram plot class supports may be a major pain to implement. Sklearn itself only supports matplotlib dendrograms as outputs
- [ ] The "Plot: Clustergram" step and HAC cannot take input from a t-test
- [ ] The "Plot: Clustergram" step and HAC cannot deal with NaNs

### LINKED ISSUE #222: t-test cannot take input directly from an imputation step
## Issue description
When setting an imputation step as an input, the t-test does not correctly select the output protein_df of the imputation step as its input. You need to add another step in between.

Note: This might very well be related to the general broken-ness of the way we obtain inputs from steps that will be fixed with #179 and #186 

## Acceptance Criteria
- [ ] The input for the t-test is always the output from the step selected in the drop-down, nothing else.

### PR DESCRIPTION
**Pull request by 3dot141592**
## Description
fixed discrepancies between the AML paper and ProtZilla.

This PR aligns ProtZilla’s proteomics pipeline with the preprocessing and statistical workflow used in the reference paper. It fixes the previously observed mismatch in protein counts and “significant proteins” by correcting the width adjustment implementation and by updating the t-test steps behavior (t-test combined with fold-change Z-score based filtering, with the option to run without multiple testing correction).

## Changes
Width normalization (width adjustment)
-  Fixed the scaling behavior that previously widened distributions by ~1.7× compared to the paper.

T-test
- Added support for no multiple testing correction (`none`) to reproduce the paper’s selection strategy.
- Implemented fold-change Z-score significance filtering

## Testing
1. Load the AML dataset (proteinGroups + [meta.csv](https://github.com/user-attachments/files/24740946/meta.csv)) and run the pipeline up to preprocessing.
2. Verify that after width adjustment the distributions/values match the Supplementary file.
3. Run t-test with: multiple testing correction = `none` and FC Z-score threshold = `0.08`.
4. Confirm the resulting significant protein list is close to the paper’s results (ca. 351; current reproduction yields 353 with 321 overlap vs the paper’s 351 list).

## PR checklist
**Development**
- [x] If necessary, I have updated the documentation (README, docstrings, etc.)
- [x] If necessary, I have created / updated tests.
 
**Mergeability**
- [x] main-branch has been merged into local branch to resolve conflicts
- [x] The tests and linter have passed AFTER local merge
- [x] The backend code has been formatted with `black`
- [x] The frontend code has been formatted with `pnpm format` and checked with `pnpm lint`
 
**Code review**
- [x] I have self-reviewed my code.
- [ ] At least one other developer reviewed and approved the changes

### CODE REVIEW

File: backend/protzilla/data_analysis/differential_expression_helper.py [hunk 1]
```diff
@@ -15,21 +15,25 @@ def apply_multiple_testing_correction(
...
-        Can be either "Bonferroni" or "Benjamini-Hochberg"
+        Can be either "Bonferroni", "Benjamini-Hochberg", or "None"
     :param alpha: the alpha value to be used for the correction
     :return: a tuple containing the corrected p-values and (depending on the correction method)\
           either the input alpha value or the corrected alpha value
     """
     assert method in [
         "Bonferroni",
         "Benjamini-Hochberg",
+        "None",
```
Comment (tE3m) [inline backend/protzilla/data_analysis/differential_expression_helper.py]: why not check against the enum defined for the form? e.g.
```suggestion
assert method in [option.value for option in MultipleTestingCorrectionMethod
```
this would of course require the import of that enum

**Comment by tE3m** [comment]
I'm not certain about the concrete steps to run for testing. Currently, I'm using this:
- Importing
  - Protein Groups Import
  - Metadata Import
- Preprocessing
  - Inversion
  - Log
    - base 2
  - Width adjustment
    - Group by None?
  - SILAC-ratios
    - 5
- Data Analysis
  - t-test
    - Welch's
    - FilterProteinsBySilacRatios_1
    - None
    - 0.05
    - Group
    - REL-FREE
    - RELAPSE
    - box ticked
    - 0.08

is this what's intended?

**Comment by 3dot141592** [comment]
> I'm not certain about the concrete steps to run for testing. Currently, I'm using this:
> 
>     * Importing
>       
>       * Protein Groups Import
>       * Metadata Import
> 
>     * Preprocessing
>       
>       * Inversion
>       * Log
>         
>         * base 2
>       * Width adjustment
>         
>         * Group by None?
>       * SILAC-ratios
>         
>         * 5
> 
>     * Data Analysis
>       
>       * t-test
>         
>         * Welch's
>         * FilterProteinsBySilacRatios_1
>         * None
>         * 0.05
>         * Group
>         * REL-FREE
>         * RELAPSE
>         * box ticked
>         * 0.08
> 
> 
> is this what's intended?

@tE3m  @NeleRiediger 

First of all: Sorry for the completely inadequate test-description. I wanted to describe it more extensive, but I lost sight of it.

In the beginning of the Preprocessing step a 'SILAC-ratios' with `1` is needed as well - The rest of your workflow looks fine tarek.

You can also find the exported workflow on our [Nextcloud](https://nextcloud.hpi.de/apps/files/files?dir=/PROTzilla_2025).

You can use the following analyzer for identifying how many different unique proteins are in your exported data.csv: [analyzer.zip](https://github.com/user-attachments/files/25017987/analyzer.zip) (replace your exportes data.csv file and run the analyzer.py. You might need to install openpyxl or pandas)

Again: Im sorry this comment is coming this late.

File: backend/protzilla/data_analysis/differential_expression_helper.py [hunk 2]
```diff
@@ -15,21 +15,25 @@ def apply_multiple_testing_correction(
...
     ], "Invalid multiple testing correction method"
     assert all(
         isinstance(i, (int, float)) and not math.isnan(i) and i is not None
         for i in p_values
     ), "List contains non-number or NaN values"
     assert 0 <= alpha <= 1, "Alpha value must be between 0 and 1"
 
+    if method == "None":
```
Comment (tE3m) [inline backend/protzilla/data_analysis/differential_expression_helper.py]: here, the enum could be used as well
```suggestion
    if method == MultipleTestingCorrectionMethod.none:
```

File: backend/protzilla/methods/data_analysis.py [hunk 3]
```diff
@@ -1120,7 +1138,14 @@ def modify_form(self, form, run):
             form_helper.get_choices_for_protein_df_steps(
                 run,
             )
+            + form_helper.to_choices(
+                run.steps.get_instance_identifiers(
+                    Step,
+                    ["significant_proteins_df"],
```
Comment (tE3m) [inline backend/protzilla/methods/data_analysis.py]: no reason for this to be a list with length 1
```suggestion
                    "significant_proteins_df",
```

**Review by tE3m — COMMENTED** [review]
I reproduced the workflow to the same result.
Only some code-style comments.

File: backend/protzilla/data_analysis/differential_expression_t_test.py [hunk 4]
```diff
@@ -28,6 +28,8 @@ def t_test(
     alpha: float,
     log_base: str = None,
     intensity_name: str = None,
+    fc_zscore_filter: bool = False,
+    fc_zscore_alpha: float = 0.05,
```
Comment (NeleRiediger) [inline backend/protzilla/data_analysis/differential_expression_t_test.py:38]: Is it clean to use magic numbers like this? Don't know if there is any other sane option to do it, but maybe just add a comment indicating "where the value comes from"

File: backend/protzilla/data_analysis/differential_expression_t_test.py [hunk 5]
```diff
@@ -158,10 +166,27 @@ def t_test(
             log2_fold_change_df=pd.DataFrame(
                 columns=["Protein ID", "log2_fold_change"]
             ),
+            fc_significance_df=pd.DataFrame(
+                columns=["Protein ID", "fc_z_score", "fc_significance"]
+            ),
```
Comment (NeleRiediger) [inline backend/protzilla/data_analysis/differential_expression_t_test.py]: Those 3 lines are the same as 105-107, also this exact list of columns is used one or two more times. Maybe there is a good way to outsource this, to make eventual refactoring in the future easier?

File: backend/protzilla/data_analysis/differential_expression_t_test.py [hunk 6]
```diff
@@ -197,13 +223,21 @@ def t_test(
...
         differentially_expressed_proteins_df=differentially_expressed_proteins_df,
         significant_proteins_df=significant_proteins_df,
         corrected_p_values_df=corrected_p_values_df,
         t_statistic_df=t_statistic_df,
         log2_fold_change_df=log2_fold_change_df,
+        fc_significance_df=fc_significance_df,
         corrected_alpha=corrected_alpha,
+        fc_zscore_alpha=fc_zscore_alpha,
+        fc_zscore_filter=fc_zscore_filter,
+        # filtered_proteins=filtered_proteins,
```
Comment (NeleRiediger) [inline backend/protzilla/data_analysis/differential_expression_t_test.py]: Should this commented line be deleted?

**Review by NeleRiediger — COMMENTED** [review]
Looks good to me too. Apart from the comments that were already made and a few small comments from my side

File: backend/protzilla/data_analysis/differential_expression_t_test.py (hunk 4 above)
Comment (tE3m) [reply backend/protzilla/data_analysis/differential_expression_t_test.py:38]: My take: for default values, I think this is fine. This method will only ever be called after the form is filled out, where this field is required ad the default value is the same

File: backend/protzilla/data_analysis/differential_expression_helper.py (hunk 2 above)
Comment (tE3m) [reply backend/protzilla/data_analysis/differential_expression_helper.py]: I still feel this is better than hardcoding the value

**Review by tE3m — COMMENTED** [review]
Looks good, but take a look at my unresolved comments. If you feel differently, feel free to reply/close them.

File: backend/protzilla/data_analysis/differential_expression_t_test.py (hunk 5 above)
Comment (3dot141592) [reply backend/protzilla/data_analysis/differential_expression_t_test.py]: done :)

File: backend/protzilla/data_analysis/differential_expression_t_test.py (hunk 6 above)
Comment (3dot141592) [reply backend/protzilla/data_analysis/differential_expression_t_test.py]: yep, thank you :). Done
